curve-amm/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+// CurveAmm Dex
+// Implementation based on https://github.com/equilibrium-eosdt/equilibrium-curve-amm/blob/master/pallets/equilibrium-curve-amm/src/lib.rs
+// For more details read
+//      https://miguelmota.com/blog/understanding-stableswap-curve/
+//      https://curve.fi/files/stableswap-paper.pdf
+//      https://github.com/equilibrium-eosdt/equilibrium-curve-amm/blob/master/docs/deducing-get_y-formulas.pdf
+
+#![no_std]
+#![feature(const_btree_new)]
+
+extern crate alloc;
+
+use alloc::collections::{BTreeMap, BTreeSet};
+use alloc::{vec, vec::Vec};
+use codec::{Decode, Encode};
+use core::num::ParseIntError;
+use fungible_token_messages::{
+    Action, BurnInput, Event, MintInput, TransferFromInput, TransferInput,
+};
+use gstd::{errors::ContractError, exec, lock::mutex::Mutex, msg, prelude::*, ActorId, ToString};
+use scale_info::TypeInfo;
+use sp_arithmetic::{
+    fixed_point::FixedU128,
+    per_things::Permill,
+    traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Zero},
+    FixedPointNumber,
+};
+
+fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
+    (0..s.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&s[i..i + 2], 16))
+        .collect()
+}
+
+/// Return Err of the expression: `return Err($expression);`.
+///
+/// Used as `fail!(expression)`.
+#[macro_export]
+macro_rules! fail {
+    ( $y:expr ) => {{
+        return Err($y.into());
+    }};
+}
+
+/// Evaluate `$x:expr` and if not true return `Err($y:expr)`.
+///
+/// Used as `ensure!(expression_to_ensure, expression_to_return_on_false)`.
+#[macro_export]
+macro_rules! ensure {
+    ( $x:expr, $y:expr $(,)? ) => {{
+        if !$x {
+            $crate::fail!($y);
+        }
+    }};
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+struct CurveAmmInitConfig {
+    /// token accounts
+    token_accounts: Vec<u8>,
+    /// amplification_coefficient
+    amplification_coefficient: u128,
+    /// fee
+    fee: u32,
+    /// admin fee
+    admin_fee: u32,
```

Comment:
As @dkubatko noted earlier we have to either add meaningful doc-comments (and according to Rust-doc style) or remote senseless comments.